<a href="https://colab.research.google.com/github/bernieee/COURSERA/blob/master/Intro_to_ML/week7_dota2_game.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [206]:
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import roc_auc_score
import time
import datetime
from random import randint
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

In [ ]:
!unzip features.zip 
!unzip features_test.zip

Archive:  features.zip
  inflating: features.csv            
Archive:  features_test.zip
  inflating: features_test.csv       


## Data

In [177]:
features = pd.read_csv('features.csv', index_col='match_id')

features.head()

start_time  lobby_type  ...  barracks_status_radiant  barracks_status_dire
match_id                          ...                                               
0         1430198770           7  ...                       51                     0
1         1430220345           0  ...                       63                     1
2         1430227081           7  ...                        0                    63
3         1430263531           1  ...                       50                    63
4         1430282290           7  ...                        3                    63

[5 rows x 108 columns]

In [ ]:
y = features['radiant_win']

In [178]:
features = features.drop(features.columns[102:], axis=1)

features.head()

start_time  lobby_type  ...  dire_ward_sentry_count  dire_first_ward_time
match_id                          ...                                              
0         1430198770           7  ...                       2                 -52.0
1         1430220345           0  ...                       1                  -5.0
2         1430227081           7  ...                       1                  13.0
3         1430263531           1  ...                       0                  27.0
4         1430282290           7  ...                       0                 -16.0

[5 rows x 102 columns]

In [ ]:
features.shape

(97230, 102)

In [179]:
features[features.columns[0]].count()

97230

In [ ]:
num_of_values = features.shape[0]

for column in features.columns:
    if features[column].count() < num_of_values:
        print(column, '=', features[column].count())

first_blood_time = 77677
first_blood_team = 77677
first_blood_player1 = 77677
first_blood_player2 = 53243
radiant_bottle_time = 81539
radiant_courier_time = 96538
radiant_flying_courier_time = 69751
radiant_first_ward_time = 95394
dire_bottle_time = 81087
dire_courier_time = 96554
dire_flying_courier_time = 71132
dire_first_ward_time = 95404


In [180]:
features.fillna(0, inplace=True)

In [ ]:
X = features.values
y = y.values

In [ ]:
X

array([[ 1.43019877e+09,  7.00000000e+00,  1.10000000e+01, ...,
         2.00000000e+00,  2.00000000e+00, -5.20000000e+01],
       [ 1.43022034e+09,  0.00000000e+00,  4.20000000e+01, ...,
         3.00000000e+00,  1.00000000e+00, -5.00000000e+00],
       [ 1.43022708e+09,  7.00000000e+00,  3.30000000e+01, ...,
         3.00000000e+00,  1.00000000e+00,  1.30000000e+01],
       ...,
       [ 1.45029185e+09,  1.00000000e+00,  9.80000000e+01, ...,
         3.00000000e+00,  1.00000000e+00, -1.50000000e+01],
       [ 1.45029299e+09,  1.00000000e+00,  1.00000000e+02, ...,
         3.00000000e+00,  1.00000000e+00, -4.20000000e+01],
       [ 1.45031337e+09,  7.00000000e+00,  5.00000000e+01, ...,
         2.00000000e+00,  0.00000000e+00, -3.90000000e+01]])

## Cross-validation

In [ ]:
folds = KFold(n_splits=5, shuffle=True)

for train_id, test_id in folds.split(X):
    X_train, X_test = X[train_id], X[test_id]
    y_train, y_test = y[train_id], y[test_id] 


In [ ]:
X_train.shape
y

(77784, 102)

## Gradient Boosting

In [ ]:
start_time = datetime.datetime.now()

clf = GradientBoostingClassifier(n_estimators=30)
clf.fit(X_train, y_train)

print('Time elapsed:', datetime.datetime.now() - start_time)

y_pred = clf.predict_proba(X_test)[:, 1]

score = roc_auc_score(y_test, y_pred)
print(score)Time elapsed: 0:00:26.917071
0.6874905149754025

Time elapsed: 0:00:26.917071
0.6874905149754025


In [ ]:
start_time = datetime.datetime.now()

clf = GradientBoostingClassifier(n_estimators=100)
clf.fit(X_train, y_train)

print('Time elapsed:', datetime.datetime.now() - start_time)

y_pred = clf.predict_proba(X_test)[:, 1]

score = roc_auc_score(y_test, y_pred)
print(score)

Time elapsed: 0:02:01.466015
0.70632141612506


In [ ]:
start_time = datetime.datetime.now()

clf = GradientBoostingClassifier(n_estimators=100, max_depth=6)
clf.fit(X_train, y_train)

print('Time elapsed:', datetime.datetime.now() - start_time)

y_pred = clf.predict_proba(X_test)[:, 1]

score = roc_auc_score(y_test, y_pred)
print(score)

Time elapsed: 0:04:15.155510
0.715950480702828


In [ ]:
start_time = datetime.datetime.now()

clf = GradientBoostingClassifier(n_estimators=1000, max_depth=6)
clf.fit(X_train, y_train)

print('Time elapsed:', datetime.datetime.now() - start_time)

y_pred = clf.predict_proba(X_test)[:, 1]

score = roc_auc_score(y_test, y_pred)
print(score)

Time elapsed: 0:30:52.383738
0.7132874462588823


In [ ]:
start_time = datetime.datetime.now()

clf = GradientBoostingClassifier(n_estimators=30, max_depth=6)
clf.fit(X_train, y_train)

print('Time elapsed:', datetime.datetime.now() - start_time)

y_pred = clf.predict_proba(X_test)[:, 1]

score = roc_auc_score(y_test, y_pred)
print(score)

Time elapsed: 0:01:12.956606
0.7021680889212464


In [ ]:
start_time = datetime.datetime.now()

clf = GradientBoostingClassifier(n_estimators=30, max_depth=randint(6, 10))
clf.fit(X_train, y_train)

print('Time elapsed:', datetime.datetime.now() - start_time)

y_pred = clf.predict_proba(X_test)[:, 1]

score = roc_auc_score(y_test, y_pred)
print(score)

Time elapsed: 0:01:13.771095
0.7022114303585422


### Anwers:

1.  Список значений с NaN:
---
    first_blood_time
    first_blood_team
    first_blood_player1
    first_blood_player2
    radiant_bottle_time
    radiant_courier_time
    radiant_flying_courier_time
    radiant_first_ward_time
    dire_bottle_time
    dire_courier_time
    dire_flying_courier_time
    dire_first_ward_time

first_blood_time -- возможно, что "первая кровь" случилась не через 5 минут после начала игры, а позже

radiant_bottle_time -- данный предмет не был преобретен

2.   radient_win -- целевое значение

3.   time = 26.917071 sec

     roc_auc_score = 0.6874905149754025

4.   При увеличении количества деревьев есть вероятность, что алгоритм переобучится. 

## Logistic Regression

#### Basic Logistic Regression

In [ ]:
scaler = StandardScaler()

X_train_scaled, X_test_scaled = scaler.fit_transform(X_train), scaler.transform(X_test)

In [ ]:
c_list = np.linspace(0.01, 0.1, num = 10, endpoint = False)
c_list

array([0.01 , 0.019, 0.028, 0.037, 0.046, 0.055, 0.064, 0.073, 0.082,
       0.091])

In [ ]:
for c in c_list:
    start_time = datetime.datetime.now()

    clf = LogisticRegression(C=c)
    clf.fit(X_train_scaled, y_train)

    print('Time elapsed:', datetime.datetime.now() - start_time)

    y_pred = clf.predict_proba(X_test_scaled)[:, 1]

    score = roc_auc_score(y_test, y_pred)
    print(score, 'c =', c)
    print('-----------------------------------------------------')

Time elapsed: 0:00:01.308795
0.7193124083854611 c = 0.01
-----------------------------------------------------
Time elapsed: 0:00:01.295782
0.7192792375127715 c = 0.019000000000000003
-----------------------------------------------------
Time elapsed: 0:00:01.279596
0.7192717261270363 c = 0.028000000000000004
-----------------------------------------------------
Time elapsed: 0:00:01.353649
0.7192662912315749 c = 0.037000000000000005
-----------------------------------------------------
Time elapsed: 0:00:01.355775
0.7192630599584449 c = 0.046000000000000006
-----------------------------------------------------
Time elapsed: 0:00:01.301980
0.7192596803645808 c = 0.05500000000000001
-----------------------------------------------------
Time elapsed: 0:00:01.287703
0.7192574767422495 c = 0.064
-----------------------------------------------------
Time elapsed: 0:00:01.320626
0.7192544891388963 c = 0.073
-----------------------------------------------------
Time elapsed: 0:00:01.288257
0.

#### Drop Categorical Features

In [176]:
columns = list(features.columns)

for i, column in enumerate(columns):
    if column == 'd5_hero':
        print(i)
        break

74


In [ ]:
drop_features = features
drop_features

In [184]:
for i in range(2, 75, 8):
    drop_features = drop_features.drop(columns[i], axis=1)

drop_features = drop_features.drop(features.columns[1], axis=1)
drop_features

start_time  r1_level  ...  dire_ward_sentry_count  dire_first_ward_time
match_id                        ...                                              
0         1430198770         5  ...                       2                 -52.0
1         1430220345         4  ...                       1                  -5.0
2         1430227081         4  ...                       1                  13.0
3         1430263531         4  ...                       0                  27.0
4         1430282290         4  ...                       0                 -16.0
...              ...       ...  ...                     ...                   ...
114402    1450265551         4  ...                       0                 -24.0
114403    1450277704         4  ...                       0                 -17.0
114404    1450291848         4  ...                       1                 -15.0
114405    1450292986         3  ...                       1                 -42.0
114406    1450313370         3  ...                       0                 -39.0

[97230 rows x 91 columns]

In [185]:
X_no_hero = drop_features.values
y_no_hero = y

In [186]:
for train_id, test_id in folds.split(X):
    X_train_no_hero, X_test_no_hero = X_no_hero[train_id], X_no_hero[test_id]
    y_train_no_hero, y_test_no_hero = y_no_hero[train_id], y_no_hero[test_id]

In [187]:
X_train_no_hero_scaled, X_test_no_hero_scaled = scaler.fit_transform(X_train_no_hero), scaler.transform(X_test_no_hero)

In [189]:
c_list = np.linspace(0.1, 5, num = 10, endpoint = False)
c_list

array([0.1 , 0.59, 1.08, 1.57, 2.06, 2.55, 3.04, 3.53, 4.02, 4.51])

In [190]:
for c in c_list:
    start_time = datetime.datetime.now()

    clf = LogisticRegression(C=c)
    clf.fit(X_train_no_hero_scaled, y_train_no_hero)

    print('Time elapsed:', datetime.datetime.now() - start_time)

    y_pred = clf.predict_proba(X_test_no_hero_scaled)[:, 1]

    score = roc_auc_score(y_test_no_hero, y_pred)
    print(score, 'c =', c)
    print('-----------------------------------------------------')

Time elapsed: 0:00:01.227463
0.7170410798063398 c = 0.1
-----------------------------------------------------
Time elapsed: 0:00:01.205612
0.7170344186809388 c = 0.5900000000000001
-----------------------------------------------------
Time elapsed: 0:00:01.208466
0.7170338044595186 c = 1.08
-----------------------------------------------------
Time elapsed: 0:00:01.146390
0.7170333808585393 c = 1.5700000000000003
-----------------------------------------------------
Time elapsed: 0:00:01.175725
0.7170334443986861 c = 2.06
-----------------------------------------------------
Time elapsed: 0:00:01.180689
0.7170333490884657 c = 2.5500000000000003
-----------------------------------------------------
Time elapsed: 0:00:01.184027
0.7170334549887106 c = 3.0400000000000005
-----------------------------------------------------
Time elapsed: 0:00:01.200236
0.7170332855483188 c = 3.5300000000000002
-----------------------------------------------------
Time elapsed: 0:00:01.222321
0.717033338498

#### Unique Heros

In [207]:
hero_list = np.concatenate((features['r1_hero'], features['r2_hero'], features['r3_hero'], features['r4_hero'], features['r5_hero'],
                           features['d1_hero'], features['d2_hero'], features['d3_hero'], features['d4_hero'], features['d5_hero']))

hero_list_unique = set(hero_list)
num_of_heros = max(hero_list_unique)
print(hero_list_unique)

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 109, 110, 112}


#### Bag Of Words

In [210]:
X_pick = np.zeros((features.shape[0], num_of_heros))

for i,  match_id in enumerate(features.index):
    for p in range(5):
        X_pick[i, features.loc[match_id, 'r%d_hero' % (p + 1)] - 1] = 1
        X_pick[i, features.loc[match_id, 'd%d_hero' % (p + 1)] - 1] = -1

In [216]:
X_pick.shape

(97230, 112)

In [215]:
drop_features_bag = drop_features

In [218]:
drop_features_bag = drop_features_bag.join(pd.DataFrame(X_pick, index=drop_features_bag .index)) 
drop_features_bag

start_time  r1_level  r1_xp  r1_gold  r1_lh  ...  107  108  109  110  111
match_id                                               ...                         
0         1430198770         5   2098     1489     20  ...  0.0  0.0  0.0  0.0  0.0
1         1430220345         4   1188     1033      9  ...  0.0  0.0  0.0  0.0  0.0
2         1430227081         4   1319     1270     22  ...  0.0  0.0  0.0  0.0  0.0
3         1430263531         4   1779     1056     14  ...  0.0  0.0  0.0  0.0  0.0
4         1430282290         4   1431     1090      8  ...  0.0  0.0  0.0  0.0  0.0
...              ...       ...    ...      ...    ...  ...  ...  ...  ...  ...  ...
114402    1450265551         4   1706     1198     17  ...  0.0  0.0  0.0  0.0  0.0
114403    1450277704         4   1793     1416     17  ...  0.0  0.0  1.0  0.0  0.0
114404    1450291848         4   1399      540      1  ...  0.0  0.0  0.0  0.0  1.0
114405    1450292986         3   1135      766      6  ...  0.0  0.0  0.0  0.0  0.0
114406    1450313370         3   1053      799      7  ...  0.0  0.0  0.0  0.0 -1.0

[97230 rows x 203 columns]

In [219]:
X_bag = drop_features_bag.values
y_bag = y

In [220]:
for train_id, test_id in folds.split(X_bag):
    X_train_bag, X_test_bag = X_bag[train_id], X_bag[test_id]
    y_train_bag, y_test_bag = y_bag[train_id], y_bag[test_id] 


In [221]:
X_train_bag_scaled, X_test_bag_scaled = scaler.fit_transform(X_train_bag), scaler.transform(X_test_bag)

In [226]:
c = 0.1
start_time = datetime.datetime.now()

clf = LogisticRegression(C=c)
clf.fit(X_train_bag_scaled, y_train_bag)

print('Time elapsed:', datetime.datetime.now() - start_time)

y_pred = clf.predict_proba(X_test_bag_scaled)[:, 1]

score = roc_auc_score(y_test_bag, y_pred)
print(score, 'c =', c)
print('max =', max(y_pred))
print('min =', min(y_pred))

Time elapsed: 0:00:02.188786
0.7519428802241966 c = 0.1
max = 0.9981083081549001
min = 0.008052054513664104


### Anwers:

1.  roc_auc_score = 0.7193124083854611

Качество логистической регрессии немного лучше, чем качество градиентного бустинга потому что в градиентном бустинге используется нескольно простых деревьев, которые не всегда могут хорошо приблизить вектор антиградиента, поэтому больше вероятность ошибки.
Логистическая регрессия работает быстрее.

2.   roc_auc_score = 0.7170410798063398

Ухудшение качества, возможно, происходит из-за того, что выигрыш в игре зависит от набора персонажей в команде и их характеристик, следовательно, выкидывание данных признаков, ухудшает качество предсказание. 

3.   108

4.   roc_auc_score = 0.7519428802241966

Качество улучшилось потому что то, какой игрок выбран в команду сильно влияет на дальнейший ход игры, и при использовании мешка слов мы фактически делаем из категориальных признаков-игроков бинарные признаки для каждого из уникальных персонажей, тем самым упрощаем работу алгоритму.

5.  max = 0.9981083081549001

    min = 0.008052054513664104